Add daily captcha for Javalab demo mode

bencodeorg opened the pull request:
Adds a captcha prompt for unverified teachers to complete once every 24 hours. Teachers are prompted once they hit the run button -- they have to hit run again to execute their code once they have verified. Added console messages to clarify when a user has completed their verification.

<img width="674" alt="image" src="https://user-images.githubusercontent.com/25372625/236886650-8395a45f-ac73-4246-a99e-3c4a5844d0b6.png">

Changes here include:
- Modified existing (but unused) `ReCaptchaDialog` component to support this use case. Opening/closing the dialog required some rework to rerender the captcha each time it was opened.
- New serialized attribute on User model to keep track of the last time a user has answered a captcha.

## Links

- jira ticket: [SL-813](https://codedotorg.atlassian.net/browse/SL-813)

## Testing story

Tested manually for an unverified teacher account (see video -- off-screen, I reset my account captcha status mid-video so I could show the dialog when the test button was clicked):
- gets prompted the first time they try and run code
- does not get prompted once they have submitted a captcha (and connects to javabuilder-demo)
- gets re-prompted if they open the dialog and close it without submitting the captcha

https://user-images.githubusercontent.com/25372625/236887114-4b467e78-5479-4898-8613-23f67b5a8740.mp4

We should see no other behavior change for other user types -- also tested manually that a verified teacher connected to main javabuilder unprompted, but relying on automated tests otherwise.

Also added unit tests to cover some of the new behavior.

## Deployment strategy

This PR is branched off of the initial work to point unverified teachers to our javabuilder-demo stack ([PR here](https://github.com/code-dot-org/code-dot-org/pull/51554)) -- once this has been reviewed, I'll merge the two bits of work together and we can merge them into staging together when we're ready.

bencodeorg commented on dashboard/app/models/user.rb:
```diff
@@ -2453,7 +2454,7 @@ def reset_section_attempts?
   end
 
-  def display_captcha?
+  def display_join_section_captcha?
```
Renamed this to disambiguate from this new feature. Part of some unfinished work from 2+ years ago from someone who no longer works here to prevent spam joining of sections, further documented here:

https://codedotorg.atlassian.net/browse/LP-1717

bencodeorg commented on apps/test/unit/javalab/JavalabCaptchaDialogTest.js:
```diff
@@ -0,0 +1,62 @@
+    fetchSpy;
+
+  beforeEach(() => {
+    fetchSpy = sinon.stub(window, 'fetch');
```
Wasted a lot of time before I realized it doesn't seem like you can stub the fetch API using sinon's fakeServer...not sure if there's a better way to do this, but this is what I saw in a few tests that have done something similar.

Also, if people have better ideas on how to test this dialog than what I'm doing (directly invoking a prop, which I know is a bit of an anti-pattern), let me know.

sanchitmalhotra126 replied:
This seems fine to me? You're trying to test the functionality of the onSubmit function, and I don't know if there's any better way than just grabbing it from the props of the child component.

bencodeorg replied:
Yeah I guess I've heard something like "do the UI action (eg, click on something) and then assert on behavior", but not really possible in this case unless I want to mount the whole component. Since this component is really just a wrapper, maybe it makes sense that I'm asserting directly on the prop.

molly-moen commented on apps/src/javalab/JavalabCaptchaDialog.jsx:
```diff
@@ -0,0 +1,76 @@
+    }).then(() => {
+      appendOutputLog(javalabMsg.verificationSuccessful());
+      appendNewlineToConsoleLog();
+      setIsCaptchaDialogOpen(false);
```
can we auto-run the program here rather than making the user click run again? (We would probably need to make sure we tracked whether they clicked 'run' or 'test' and trigger the correct type again)

bencodeorg replied:
Yeah I wasn't sure about this without a product spec, seemed a little tricky so went for the simpler version first. Let me take a stab and see if I can whip something up!

molly-moen reviewed (commented):
This looks good! Questions/thoughts:
- The video showed a red message in the captcha about it only being for testing purposes. Should we remove that now?
- I think a more seamless user flow would be after getting the captcha verified, we automatically run their program without them needing to click again. I added a comment where I think we could trigger this.

bencodeorg commented:
> * The video showed a red message in the captcha about it only being for testing purposes. Should we remove that now?

Sorry should have clarified that, that "testing purposes" message depends on which key you're using (I'm using our test key locally). I can witch to our production key briefly just to double check that message disappears.

bencodeorg commented:
@molly-moen updated to continue execution after submitting the captcha, and updated screenshots / video to use production keys (now deleted!) so you can see what it would actually look like. Take a look!

bencodeorg commented on apps/src/javalab/JavalabView.jsx:
```diff
@@ -140,6 +141,26 @@ class JavalabView extends React.Component {
+  };
+
+  onCaptchaCancel = () => {
+    const {isRunning, isTesting, setIsRunning, setIsTesting} = this.props;
```
This is...not beautiful, but I feel like our run/test state is unnecessarily complicated? Should we make this an enum or something?

molly-moen replied:
hmm yeah if you wanted to we could change from flags to a single enum `RUNNING, TESTING, OFF` or something like that? But that doesn't have to happen as part of this work, as it probably requires changing a few files

jmkulwik replied:
Similarly, if we see ourselves doing this in a couple places, it'd be worth setting up some sort of (locally) global setter function so we're guaranteed to never have more than one of these states set to true at a time.

bencodeorg replied:
Yeah agree with both of these points -- I just added an item here to look into this more:

https://codedotorg.atlassian.net/browse/SL-852

molly-moen commented on apps/src/templates/ReCaptchaDialog.jsx:
```diff
@@ -32,19 +34,44 @@ export default class ReCaptchaDialog extends React.Component {
+
+  // The dialog is not fully unmounted when isOpen is set to false,
+  // but the captcha is removed from the DOM.
+  // Thus, we need to force re-rerender the captcha each time the dialog is opened.
```
nit: should this say `rerender`?

molly-moen reviewed (approved):
LGTM!

jmkulwik commented on apps/src/javalab/Javalab.js:
```diff
@@ -192,6 +193,7 @@ Javalab.prototype.init = function (config) {
     isResponsive: true,
     isSubmittable: !!config.level.submittable,
     isSubmitted: !!config.level.submitted,
+    recaptchaSiteKey: config.level.recaptchaSiteKey,
```
Do we have any concerns about someone grabbing this key and spamming the captcha endpoint?

bencodeorg replied:
It's already accessible if you're reporting abuse, I believe this is standard practice for captchas? Some docs [here](https://developers.google.com/recaptcha/intro)

https://studio.code.org/report_abuse

sanchitmalhotra126 commented on apps/src/javalab/JavalabCaptchaDialog.jsx:
```diff
@@ -0,0 +1,86 @@
+  recaptchaSiteKey: PropTypes.string.isRequired,
+};
+
+export default connect(
```
This might make testing more complicated, so not necessary, but you could use [redux hooks](https://react-redux.js.org/api/hooks) in this component and remove all the `connect` and redux props boilerplate

sanchitmalhotra126 commented on apps/src/javalab/JavalabCaptchaDialog.jsx:
```diff
@@ -0,0 +1,86 @@
+}) {
+  // return promise for tests
+  const onCaptchaSubmit = token => {
+    return fetch('/dashboardapi/v1/users/me/verify_captcha', {
```
Just curious - it's interesting that the `RecaptchaDialog` itself doesn't already just make this call? Do you know if we have other/multiple recaptcha endpoints?

bencodeorg replied:
Yeah the `RecaptchaDialog` is not in use at the moment (and explicitly tells you to make your own server call). I think the more standard way to use captchas is in a form, which is what we do at studio.code.org/report_abuse:

https://github.com/code-dot-org/code-dot-org/blob/5a7df81de4108acf7158ce2339516dfc43e95422/apps/src/code-studio/components/ReportAbuseForm.jsx#L169-L175

Google loads the captcha into the form (wherever the special `g-recaptcha` class name is) -- it submits a google generated token as part of the form submission, which can be validated in a controller using this magic `verify_captcha` method:

https://github.com/code-dot-org/code-dot-org/blob/5a7df81de4108acf7158ce2339516dfc43e95422/dashboard/app/controllers/report_abuse_controller.rb#L37-L41

So, we definitely use captcha in a one or a couple other places, but they're typically less about an explicit captcha verification (and storage of a timestamp when verification happened) and more used to guard access to some endpoint.

bencodeorg commented:
Closing in favor of:

https://github.com/code-dot-org/code-dot-org/pull/51886
